csi/node.go

```diff
@@ -30,3 +52,128 @@ func (s *OsdCsiServer) GetNodeID(ctx context.Context, req *csi.GetNodeIDRequest)
 
 	return result, nil
 }
+
+// NodePublishVolume is a CSI API call which mounts the volume on the specified
+// target path on the node.
+//
+// TODO: Support READ ONLY Mounts
+//
+func (s *OsdCsiServer) NodePublishVolume(
+	ctx context.Context,
+	req *csi.NodePublishVolumeRequest,
+) (*csi.NodePublishVolumeResponse, error) {
+
+	dlog.Debugf("NodePublishVolume req[%#v]", req)
+
+	// Check arguments
+	if req.GetVersion() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Version must be provided")
+	}
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume id must be provided")
+	}
+	if len(req.GetTargetPath()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Target path must be provided")
+	}
+	if req.GetVolumeCapability() == nil || req.GetVolumeCapability().GetAccessMode() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Volume access mode must be provided")
+	}
+
+	// Get volume information
+	v, err := util.VolumeFromName(s.driver, req.GetVolumeId())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Volume id %s not found: %s",
+			req.GetVolumeId(),
+			err.Error())
+	}
+
+	// Gather volume attributes
+	spec, _, _, err := s.specHandler.SpecFromOpts(req.GetVolumeAttributes())
+	if err != nil {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"Invalid volume attributes: %#v",
+			req.GetVolumeAttributes())
+	}
+
+	// This seems weird as a way to change opts to map[string]string
+	opts := make(map[string]string)
+	if len(spec.GetPassphrase()) != 0 {
+		opts[options.OptionsSecret] = spec.GetPassphrase()
+	}
+
+	// Verify target location is an existing directory
+	// See: https://github.com/container-storage-interface/spec/issues/60
+	err = verifyTargetLocation(req.GetTargetPath())
+	if err != nil {
+		return nil, status.Error(codes.NotFound, err.Error())
+	}
+
+	// If this is for a block driver, first attach the volume
+	if s.driver.Type() == api.DriverType_DRIVER_TYPE_BLOCK {
+		_, err := s.driver.Attach(req.GetVolumeId(), opts)
+		if err != nil {
+			return nil, status.Errorf(
+				codes.Internal,
+				"Unable to attach volume: %s",
+				err.Error())
+		}
+	}
+
+	// Mount volume onto the path
+	err = s.driver.Mount(v.GetId(), req.GetTargetPath(), nil)
+	if err != nil {
+		// Detach on error
+		detachErr := s.driver.Detach(v.GetId(), opts)
+		if detachErr != nil {
+			dlog.Errorf("Unable to deatch volume %s: %s",
+				v.GetId(),
+				detachErr.Error())
+		}
+		return nil, status.Errorf(
+			codes.Internal,
+			"Unable to mount volume %s onto %s: %s",
+			req.GetVolumeId(),
+			req.GetTargetPath(),
+			err.Error())
+	}
+
+	dlog.Infof("Voume %s mounted on %s",
+		req.GetVolumeId(),
+		req.GetTargetPath())
+
+	return &csi.NodePublishVolumeResponse{}, nil
+}
+
+func verifyTargetLocation(targetPath string) error {
+	fileinfo, err := os.Stat(targetPath)
+	if err != nil && os.IsNotExist(err) {
+		return status.Errorf(
+			codes.Aborted,
```

Comment:
Notfound is more for the request which was to mount a volume. NotFound means the volume is not found. I think aborted is more applicable here since the command was aborted due to an error. I could also use code.InvalidArgument since the caller provided this argument and is invalid.

Reply:
I am okay with not sending Notfound here as I agree that it is not the volume that is not found. Aborted is also fine with me.
Just in that case, make sure at the top where you call this, you don't return NotFound.